File: core/src/test/java/xtdb/bitemporal/PolygonTest.java
package xtdb.bitemporal;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

class PolygonTest {

    private static final long MAX_LONG = Long.MAX_VALUE;

    private Polygon polygon;
    private Polygon inPolygon;
    private Ceiling ceiling;

    @BeforeEach
    void setUp() {
        polygon = new Polygon();

        inPolygon = new Polygon();
        inPolygon.sysTimeCeilings().add(MAX_LONG);

        ceiling = new Ceiling();
    }

    void setInValidTimes(long... validTimes) {
        inPolygon.validTimes().clear();
        inPolygon.validTimes().add(validTimes);
    }

    void setInSysTimeCeilings(long... sysTimeCeilings) {
        inPolygon.sysTimeCeilings().clear();
        inPolygon.sysTimeCeilings().add(sysTimeCeilings);
    }

    void applyPolygon(long sysFrom) {
        polygon.calculateFor(ceiling, inPolygon);
        ceiling.applyLog(sysFrom, inPolygon.getValidFrom(0), inPolygon.getValidTo(inPolygon.getValidTimeRangeCount() - 1));
    }

    void applyEvent(long sysFrom, long validFrom, long validTo) {
        setInValidTimes(validFrom, validTo);
        applyPolygon(sysFrom);
    }

    void assertValidTimesAre(long... validTimes) {
        assertArrayEquals(validTimes, polygon.validTimes().toArray());
    }

    void assertSysTimeCeilingsAre(long... sysTimeCeilings) {
        assertArrayEquals(sysTimeCeilings, polygon.sysTimeCeilings().toArray());
    }

    @Test
    void testCalculateForEmptyCeiling() {
        applyEvent(0, 2, 3);
        assertValidTimesAre(2, 3);
        assertSysTimeCeilingsAre(MAX_LONG);
    }

    @Test
    void startsBeforeNoOverlap() {
        applyEvent(1, 2005, 2009);
        assertValidTimesAre(2005, 2009);
        assertSysTimeCeilingsAre(MAX_LONG);

        applyEvent(0, 2010, 2020);
        assertValidTimesAre(2010, 2020);
        assertSysTimeCeilingsAre(MAX_LONG);
    }

    @Test
    void startsBeforeAndOverlaps() {
        applyEvent(1, 2010, 2020);
        assertValidTimesAre(2010, 2020);
        assertSysTimeCeilingsAre(MAX_LONG);

        applyEvent(0, 2015, 2025);
        assertValidTimesAre(2015, 2020, 2025);
        assertSysTimeCeilingsAre(1, MAX_LONG);
    }

    @Test
    void startsEquallyAndOverlaps() {
        applyEvent(1, 2010, 2020);
        assertValidTimesAre(2010, 2020);
        assertSysTimeCeilingsAre(MAX_LONG);

        applyEvent(0, 2010, 2025);
        assertValidTimesAre(2010, 2020, 2025);
        assertSysTimeCeilingsAre(1, MAX_LONG);
    }

    @Test
    void newerPeriodCompletelyCovered() {
        applyEvent(1, 2015, 2020);
        applyEvent(0, 2010, 2025);

        assertValidTimesAre(2010, 2015, 2020, 2025);
        assertSysTimeCeilingsAre(MAX_LONG, 1, MAX_LONG);
    }

    @Test
    void olderPeriodCompletelyCovered() {
        applyEvent(1, 2010, 2025);
        applyEvent(0, 2010, 2020);

        assertValidTimesAre(2010, 2020);
        assertSysTimeCeilingsAre(1);
    }

    @Test
    void periodEndsEquallyAndOverlaps() {
        applyEvent(1, 2015, 2025);
        applyEvent(0, 2010, 2025);

        assertValidTimesAre(2010, 2015, 2025);
        assertSysTimeCeilingsAre(MAX_LONG, 1);
    }

    @Test
    void periodEndsAfterAndOverlaps() {
        applyEvent(1, 2015, 2025);
        applyEvent(0, 2010, 2020);

        assertValidTimesAre(2010, 2015, 2020);
        assertSysTimeCeilingsAre(MAX_LONG, 1);
    }

    @Test
    void periodStartsBeforeAndTouches() {
        applyEvent(1, 2005, 2010);
        applyEvent(0, 2010, 2020);

        assertValidTimesAre(2010, 2020);
        assertSysTimeCeilingsAre(MAX_LONG);
    }

    @Test
    void periodStartsAfterAndTouches() {
        applyEvent(1, 2010, 2020);
        applyEvent(0, 2005, 2010);

        assertValidTimesAre(2005, 2010);
        assertSysTimeCeilingsAre(MAX_LONG);
    }

    @Test
    void periodStartsAfterAndDoesNotOverlap() {
        applyEvent(1, 2010, 2020);
        applyEvent(0, 2005, 2009);

        assertValidTimesAre(2005, 2009);
        assertSysTimeCeilingsAre(MAX_LONG);
    }

    @Test
    void testCalculatingFromMultiRangePolygon() {
        applyEvent(5, 2012, 2015);
        setInValidTimes(2005, 2009, 2010, Long.MAX_VALUE);
        setInSysTimeCeilings(4, 3, Long.MAX_VALUE);

        applyPolygon(1);
        assertValidTimesAre(2005, 2009, 2010, 2012, 2015, Long.MAX_VALUE);
        assertSysTimeCeilingsAre(4, 3, Long.MAX_VALUE, 5, Long.MAX_VALUE);
    }

    @Test
    void testMultiRangePolygonMeetsNewEvent() {
        applyEvent(5, 2010, 2015);
        setInValidTimes(2005, 2009, 2010, Long.MAX_VALUE);
        setInSysTimeCeilings(4, 3, Long.MAX_VALUE);

        applyPolygon(1);
        assertValidTimesAre(2005, 2009, 2010, 2015, Long.MAX_VALUE);
        assertSysTimeCeilingsAre(4, 3, 5, Long.MAX_VALUE);
    }

    @Test
    void testLaterEventDoesntChangeSupersededEvent() {
        applyEvent(5, 2010, 2015);
        setInValidTimes(2005, 2009, Long.MAX_VALUE);
        setInSysTimeCeilings(Long.MAX_VALUE, 3);

        applyPolygon(1);
        assertValidTimesAre(2005, 2009, Long.MAX_VALUE);
        assertSysTimeCeilingsAre(Long.MAX_VALUE, 3);
    }
}

File: core/src/test/kotlin/xtdb/bitemporal/PolygonTest.kt
package xtdb.bitemporal

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import com.carrotsearch.hppc.LongArrayList.from as longs
import kotlin.Long.Companion.MAX_VALUE as MAX_LONG
import kotlin.Long.Companion.MIN_VALUE as MIN_LONG

internal class PolygonTest {
    private lateinit var polygon: Polygon
    private lateinit var ceiling: Ceiling

    @BeforeEach
    fun setUp() {
        polygon = Polygon()
        ceiling = Ceiling()
    }

    private fun applyEvent(sysFrom: Long, validFrom: Long, validTo: Long) {
        polygon.calculateFor(ceiling, validFrom, validTo)

        ceiling.applyLog(
            sysFrom,
            validFrom,
            validTo
        )
    }

    @Test
    fun testCalculateForEmptyCeiling() {
        applyEvent(0, 2, 3)
        assertEquals(longs(2, 3), polygon.validTimes)
        assertEquals(longs(MAX_LONG), polygon.sysTimeCeilings)
    }

    @Test
    fun startsBeforeNoOverlap() {
        applyEvent(1, 2005, 2009)
        assertEquals(longs(2005, 2009), polygon.validTimes)
        assertEquals(longs(MAX_LONG), polygon.sysTimeCeilings)

        applyEvent(0, 2010, 2020)
        assertEquals(longs(2010, 2020), polygon.validTimes)
        assertEquals(longs(MAX_LONG), polygon.sysTimeCeilings)
    }

    @Test
    fun startsBeforeAndOverlaps() {
        applyEvent(1, 2010, 2020)
        assertEquals(longs(2010, 2020), polygon.validTimes)
        assertEquals(longs(MAX_LONG), polygon.sysTimeCeilings)

        applyEvent(0, 2015, 2025)
        assertEquals(longs(2015, 2020, 2025), polygon.validTimes)
        assertEquals(longs(1, MAX_LONG), polygon.sysTimeCeilings)
    }

    @Test
    fun startsEquallyAndOverlaps() {
        applyEvent(1, 2010, 2020)
        assertEquals(longs(2010, 2020), polygon.validTimes)
        assertEquals(longs(MAX_LONG), polygon.sysTimeCeilings)

        applyEvent(0, 2010, 2025)
        assertEquals(longs(2010, 2020, 2025), polygon.validTimes)
        assertEquals(longs(1, MAX_LONG), polygon.sysTimeCeilings)
    }

    @Test
    fun newerPeriodCompletelyCovered() {
        applyEvent(1, 2015, 2020)
        applyEvent(0, 2010, 2025)

        assertEquals(longs(2010, 2015, 2020, 2025), polygon.validTimes)
        assertEquals(longs(MAX_LONG, 1, MAX_LONG), polygon.sysTimeCeilings)
    }

    @Test
    fun olderPeriodCompletelyCovered() {
        applyEvent(1, 2010, 2025)
        applyEvent(0, 2010, 2020)

        assertEquals(longs(2010, 2020), polygon.validTimes)
        assertEquals(longs(1), polygon.sysTimeCeilings)
    }

    @Test
    fun periodEndsEquallyAndOverlaps() {
        applyEvent(1, 2015, 2025)
        applyEvent(0, 2010, 2025)

        assertEquals(longs(2010, 2015, 2025), polygon.validTimes)
        assertEquals(longs(MAX_LONG, 1), polygon.sysTimeCeilings)
    }

    @Test
    fun periodEndsAfterAndOverlaps() {
        applyEvent(1, 2015, 2025)
        applyEvent(0, 2010, 2020)

        assertEquals(longs(2010, 2015, 2020), polygon.validTimes)
        assertEquals(longs(MAX_LONG, 1), polygon.sysTimeCeilings)
    }

    @Test
    fun periodStartsBeforeAndTouches() {
        applyEvent(1, 2005, 2010)
        applyEvent(0, 2010, 2020)

        assertEquals(longs(2010, 2020), polygon.validTimes)
        assertEquals(longs(MAX_LONG), polygon.sysTimeCeilings)
    }

    @Test
    fun periodStartsAfterAndTouches() {
        applyEvent(1, 2010, 2020)
        applyEvent(0, 2005, 2010)

        assertEquals(longs(2005, 2010), polygon.validTimes)
        assertEquals(longs(MAX_LONG), polygon.sysTimeCeilings)
    }

    @Test
    fun periodStartsAfterAndDoesNotOverlap() {
        applyEvent(1, 2010, 2020)
        applyEvent(0, 2005, 2009)

        assertEquals(longs(2005, 2009), polygon.validTimes)
        assertEquals(longs(MAX_LONG), polygon.sysTimeCeilings)
    }

    @Test
    fun testTimeSeries() {
        ceiling.applyLog(10, 10, 12)
        ceiling.applyLog(8, 8, 10)
        ceiling.applyLog(6, 6, 8)

        assertEquals(longs(MAX_LONG, 12, 10, 8, 6, MIN_LONG), ceiling.validTimes)
        assertEquals(longs(MAX_LONG, 10, 8, 6, MAX_LONG), ceiling.sysTimeCeilings)

        applyEvent(4, 4, 6)
        assertEquals(longs(4, 6), polygon.validTimes)
        assertEquals(longs(MAX_LONG), polygon.sysTimeCeilings)
    }

    @Test
    fun `test single-rectangle recency`() {
        assertEquals(
            MAX_LONG,
            Polygon(longs(3, MAX_LONG), longs(MAX_LONG)).recency,
            "current"
        )

        assertEquals(
            10,
            Polygon(longs(4, 10), longs(MAX_LONG)).recency,
            "put for valid-time range"
        )

        assertEquals(
            4, Polygon(longs(6, 10), longs(4)).recency,
            "vt=tt passes above the rectangle"
        )

        assertEquals(
            6, Polygon(longs(6, 10), longs(6)).recency,
            "vt=tt touches the top-left corner"
        )

        assertEquals(
            8, Polygon(longs(6, 10), longs(8)).recency,
            "vt=tt hits the top of the rectangle"
        )

        assertEquals(
            10, Polygon(longs(6, 10), longs(10)).recency,
            "vt=tt touches the top-right corner"
        )

        assertEquals(
            10, Polygon(longs(6, 10), longs(12)).recency,
            "vt=tt hits the RHS of the rectangle"
        )
    }

    @Test
    fun `test polygon recency`() {
        assertEquals(
            5,
            Polygon(longs(3, 5, MAX_LONG), longs(MAX_LONG, 5)).recency,
            "standard vt=tt replacement"
        )

        assertEquals(
            6,
            Polygon(longs(3, 5, MAX_LONG), longs(MAX_LONG, 6)).recency,
            "retroactively corrected"
        )

        assertEquals(
            7,
            Polygon(longs(3, 7, MAX_LONG), longs(MAX_LONG, 6)).recency,
            "scheduled correction"
        )

        assertEquals(
            4,
            Polygon(longs(1, 4), longs(5)).recency,
            "historical/historical"
        )

        assertEquals(
            8,
            Polygon(longs(10, 12, 15, 18), longs(8, 6, 3)).recency,
            "sys-time descending but all with vt>tt"
        )

        assertEquals(
            8,
            Polygon(longs(10, 12, 15, 18), longs(6, 8, 3)).recency,
            "sys-time oscillating but all with vt>tt"
        )

        assertEquals(
            4,
            Polygon(longs(0, 2, 5, 8), longs(7, 4, 2)).recency,
            "sys-times descending, intersect halfway through the middle rectangle"
        )

        assertEquals(
            6,
            Polygon(longs(100, 100, 5, 8), longs(100, 9, 6)).recency,
            "sys-times descending, intersect halfway through the final rectangle - demonstrates a short circuit"
        )
    }
}
